File: Shortest_Path/c++/05_플로이드.cpp
// 이것이 취업을 위한 코딩테스트다
// https://www.acmicpc.net/problem/11404
// 1

#include <iostream>
#include <cstring>
#include <algorithm>

using namespace std;

int n, m;
int graph[101][101];

int main() {
	cin >> n >> m;

	for (int i = 0; i < 101; i++)
		fill(graph[i], graph[i] + 101, 1e9);

	for (int a = 1; a <= n; a++) {
		for (int b = 1; b <= m; b++)
			if (a == b) graph[a][b] = 0;
	}

	for (int i = 0; i < m; i++) {
		int a, b, c;
		cin >> a >> b >> c;
		if (c < graph[a][b])
			graph[a][b] = c;
	}

	for (int k = 1; k <= n; k++) {
		for (int i = 1; i <= n; i++) {
			for (int j = 1; j <= n; j++)
				graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j]);
		}
	}

	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			if (graph[i][j] == 1e9)
				cout << 0 << ' ';
			else
				cout << graph[i][j] << ' ';
		}
		cout << '\n';
	}
}

File: backup/Shortest_Path/python/05_플로이드.py
# 이것이 취업을 위한 코딩테스트다
# https://www.acmicpc.net/problem/11404
# 1

INF = int(1e9)

n = int(input())
m = int(input())
adj = [[INF] * (n + 1) for _ in range(n + 1)]

for i in range(1, n + 1):
    for j in range(1, n + 1):
        if i == j:
            adj[i][j] = 0

for _ in range(m):
    a, b, c = map(int, input().split())
    if c < adj[a][b]:
        adj[a][b] = c

for k in range(1, n + 1):
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            adj[i][j] = min(adj[i][j], adj[i][k] + adj[k][j])

for i in range(1, n + 1):
    for j in range(1, n + 1):
        if adj[i][j] == INF:
            print(0, end=' ')
        else:
            print(adj[i][j], end=' ')
    print()
